Plans after M2

Documents:

1. Report revision
2. Disease draft
3. ABACBS abstract

Research:

1. Hybrid validation: 

   1. GWAS 1000 marker
   2. Random 1000 marker 
      1. Sharing 1000 marker
      2. Individual 1000 marker subsets

2. Chromosome simulation

   1. Simulate sugarcane pseudo diploid genotype with polyploid haplotypes
   2. Remove random chromosome and regenerate diploid genotype
   3. Test accuracy changes

3. Model interpretation

   ![image-20230918150707397](C:\Users\70651\AppData\Roaming\Typora\typora-user-images\image-20230918150707397.png)

4. Bayesian neural network

5. Add Precise criteria (MSE)





Workflow:

â€‹	Hybrid model with GWAS/Random/BayesR prior knowledge

	1. create map file with marker index, save
	1. use index to create marker subset



Do GWAS learning 

```bash
composer="D:/PhD_Notebook/Code/ML_composer/"
data_path="H:/ML_archive/Data"
pheno="${data_path}/sugarcane_disease/sugarcane_disease.phen"
index="H:/ML_archive/Data/sugarcane_disease/sugarcane_disease.index"
epoch=30
loss="mse"
act="leaky_relu"
header="\tTrait\tTrainSet\tValidSet\tModel\tTest_Accuracy\tValid_Accuracy\t${loss}\tRuntime"
mkdir -p H:/ML_archive/GWAS_1000/
mkdir -p H:/ML_archive/GWAS_1000/Attention
for AB in 1
do
echo $AB
for idx in {1..5}
do
for epi in noEpi
do

if [ $epi == "Epi" ]; 
	then e="--epistatic";
elif [ $epi == "noEpi" ];
	then e="";
fi
for trait in smut pachy
do

if [ $trait == "smut" ]; 
	then tdx=1;
elif [ $trait == "pachy" ];
	then tdx=2;
fi
geno="H:/ML_archive/Data/sugarcane_disease/gwas/sugarcane_disease_gwas_1000_v${idx}_${trait}"
cp H:/ML_archive/Data/sugarcane_disease/gwas/sugarcane_disease.fam H:/ML_archive/Data/sugarcane_disease/gwas/sugarcane_disease_gwas_1000_v${idx}_${trait}.fam
target="H:/ML_archive/GWAS_1000/Attention/v${idx}_1000SNP_${epi}/"
python $composer/GS_composer.py --ped $geno --pheno $pheno --mpheno $tdx --index $index --trait $trait \
--model "MultiLevel Attention" \
--width 256 --depth 4 --locallyConnect 16 --embedding 16 --AttentionBlock $AB --num-heads 1 --locallyBlock 1 $e\
-o $target \
--quiet 1 --plot --epoch $epoch --round 1 --activation $act --batch 64 --vindex $idx --lr 0.001 --loss $loss 
done
done
done
done

########merge accuracy########
header="\tTrait\tTrainSet\tValidSet\tModel\tTest_Accuracy\tValid_Accuracy\t${loss}\tRuntime"
echo $loss
act="leaky_relu"
for AB in 1
do
for epi in noEpi
do 
echo "epi is ${epi}"
for trait in smut pachy
do
path="H:/ML_archive/GWAS_1000/Attention/"
mkdir -p $path
mkdir -p $path/${trait}_MultiLevelAttention_${trait}/
echo -e $header > $path/MultiLevelAttention_train_record_$trait.csv
for idx in {1..5}
do
echo $idx
target="H:/ML_archive/GWAS_1000/Attention/v${idx}_1000SNP/"
cat ${target}/MultiLevelAttention_train_record_${trait}.csv | grep -v "Trait" >> $path/MultiLevelAttention_train_record_${trait}.csv
cp ${target}/model_summary.txt $path/
cp  ${target}/${trait}_MultiLevelAttention_${trait}/* $path/${trait}_MultiLevelAttention_${trait}/
#echo -e $header
done
done
done
done



```



Random Modelling



```bash
composer="D:/PhD_Notebook/Code/ML_composer/"
data_path="H:/ML_archive/Data"
pheno="${data_path}/sugarcane_disease/sugarcane_disease.phen"
index="H:/ML_archive/Data/sugarcane_disease/sugarcane_disease.index"
epoch=30
loss="mse"
act="leaky_relu"
header="\tTrait\tTrainSet\tValidSet\tModel\tTest_Accuracy\tValid_Accuracy\t${loss}\tRuntime"
mkdir -p H:/ML_archive/random_1000/
mkdir -p H:/ML_archive/random_1000/Attention
for AB in 1
do
echo $AB

for epi in noEpi
do

if [ $epi == "Epi" ]; 
	then e="--epistatic";
elif [ $epi == "noEpi" ];
	then e="";
fi
for trait in smut pachy
do

if [ $trait == "smut" ]; 
	then tdx=1;
elif [ $trait == "pachy" ];
	then tdx=2;
fi
geno="H:/ML_archive/Data/sugarcane_disease/random/sugarcane_disease_random"
cp H:/ML_archive/Data/sugarcane_disease/random/sugarcane_disease.fam H:/ML_archive/Data/sugarcane_disease/random/sugarcane_disease_random.fam
target="H:/ML_archive/random_1000/Attention/random_1000SNP_${epi}/"
python $composer/GS_composer.py --ped $geno --pheno $pheno --mpheno $tdx --index $index --trait $trait \
--model "MultiLevel Attention" \
--width 256 --depth 4 --locallyConnect 16 --embedding 16 --AttentionBlock $AB --num-heads 1 --locallyBlock 1 $e\
-o $target \
--quiet 1 --plot --epoch $epoch --round 1 --activation $act --batch 64 --lr 0.001 --loss $loss 
done
done
done
done
```



Marker interface pipeline:

```bash
N: marker size
MV = data.frame()
for b in Bs
do
	pB = model.predict(Mb) # Mb shape(1,N) with all value = b;
for g in Gs
do
create_sequences M; # shape (N,N),
	M[:,:] = b
	diagonal(M) = g
make_prediction V;
	V = model.predict(M) # shape (N,)
	V$genotype = g
	colname(V)[1] = "P"
Extend V with background calculation;
	V$b = V[:,1]-pB
	V$ID = 1:N
	MV = rbind(MV,V)
	
Analysis criteria:
	Variance of allele effect:
		V$doseEff = V[:,3]/V[:,2]
		V %>% group_by(ID) %>% summarise(bVariance = var(doseEff))

```



New commend for extract snp effect from gctb

```bash
path="/scratch/user/s4563146/sugarcane_disease/all_marker/"
data_path=`realpath --relative-to=./ ./`

for trait in smut pachy
do
echo $trait
for model in BayesC BayesR
do
echo $model
for i in {1..5} 
do 
echo $i; 
cat ${model}/${trait}_v${i}.snpRes | sed 's/^[ \t]*//g' | awk -F"  +" '{print $2,$5,$8}' > ${model}/${trait}_v${i}.snpEff; 
~/plink --bfile $data_path/sugarcane_disease  --score ${model}/${trait}_v${i}.snpEff header sum --out ${model}/${trait}_v${i}.score; done
done
done

```

